qiskit_experiments/library/characterization/analysis/multi_state_discrimination_analysis.py

```diff
@@ -0,0 +1,158 @@
+        num_shots = len(experiment_data.data()[0]["memory"])
+
+        # Process the data and get labels
+        data, fit_state = [], []
+        fit_state_dict = {}
+        for i in range(n_states):
+            state_data = []
+            for j in range(num_shots):
+                state_data.append(experiment_data.data()[i]["memory"][j][0])
+            data.append(np.array(state_data))
+            label = experiment_data.data()[i]["metadata"]["label"]
+            fit_state.append(label)
+            fit_state_dict[label] = i
+
+        # Train a discriminator on the processed data
+        discriminator = self.options.discriminator
+        discriminator.fit(
+            np.concatenate(data),
+            np.asarray([[label] * num_shots for label in fit_state]).flatten().transpose(),
+        )
+
+        # Calculate fidelity. First we need to calculate P(i|j):= prob. measuring outcome i given
+        # state j was prepared
+        predicted_data = [discriminator.predict(state_data) for state_data in data]
+        # count per prepared state the number of measured states of each kind and calculate the
+        # probability of measuring the wrong state
+        prob_wrong = 0
+        for i in range(n_states):
+            counts = [0] * n_states
+            for point in predicted_data[i]:
+                counts[fit_state_dict[point]] += 1
+            for j in range(n_states):
+                if j != i:
+                    prob_wrong += counts[j] / num_shots
+
+        # calculate the fidelity
+        fidelity = 1 - (1 / n_states) * prob_wrong
+
+        # Crate analysis results from the discriminator configuration
+        analysis_results = [
+            AnalysisResultData(name="discriminator_config", value=discriminator.config()),
+            AnalysisResultData(name="fidelity", value=fidelity),
+        ]
+
+        # Create figure
+        if self.options.plot:
+            figures = [self._levels_plot(discriminator, data, fit_state, fidelity).get_figure()]
+        else:
+            figures = []
+
+        return analysis_results, figures
+
+    def _levels_plot(self, discriminator, data, fit_state, fidelity) -> matplotlib.figure.Figure:
+        """Helper function for plotting IQ plane for different energy levels.
+
+        Args:
+            discriminator: the trained discriminator
+            data: the training data
+            fit_state: the labels
+            fidelity: the fidelity of the classification
+
+        Returns:
+            The plotted IQ data.
+        """
+        # Create IQPlotter and generate figure.
+        plotter = IQPlotter(MplDrawer())
+        plotter.set_options(
+            discriminator_max_resolution=64,
+            style=PlotStyle(figsize=(6, 4), legend_loc=None),
+        )
+
+        # create figure labels
+        params_dict = {}
+        for i, label in enumerate(fit_state):
+            params_dict[label] = {"label": "$|%s\\rangle$" % i}
```

Comment:
It would be good to create an analysis-class option called `plotter` and set the default to an `IQPlotter`. This way, users can overwrite the plotter with a custom instance or customize the options before running the analysis. The plotter options and figure options should be set inside `MultiStateDiscriminationAnalysis._default_options`.

It looks like the series labels are set based on `fit_state`, which comes from the circuit metadata? Correct me if I'm wrong. If there's a way to set this in `_default_options`, that would be best.

Reply:
yes, it comes from the metadata but the circuits are given when the run method is called with the experiment data. So I'm not sure it's possible to give that as a default option.

Reply:
Got it. Then maybe the code should not overwrite `series_params`, instead query and modify `series_params` if the user hasn't set a label. The user may have set `series_params` with colours, markers, etc. and this would be overwritten by `set_figure_options`. I can suggest a change to show what I mean.

Reply:
Here's my suggestion to make the labels dynamic. It's really just one addition to make sure old `series_params` values take priority. https://github.com/Qiskit/qiskit-experiments/pull/936/files/df0b916978004c1babab7560b481a7dddf3ffade#r1016683317

It would still be good to make the plotter an option, so it is accessible outside `_level_plots`, but the aforementioned change would make labels dynamic unless set by the user. This means they don't need to be known before running the analysis, and thus set as a default option.

Reply:
Thanks! the changes are in [830be2e](https://github.com/Qiskit/qiskit-experiments/pull/936/commits/830be2e0fe6a8cfe78d9e421d882d7c56e743988). Let me know if I missed something.